xdsl/transforms/memref_stream_interleave.py

```diff
@@ -0,0 +1,154 @@
+            # Already interleaved
+            return
+
+        interleave_bound_index = -1
+        interleave_bound = -1
+        for index, (iterator_type, bound) in enumerate(
+            zip(op.iterator_types, op.bounds, strict=True)
+        ):
+            if iterator_type == memref_stream.IteratorTypeAttr.parallel():
+                interleave_bound = bound.value.data
+                interleave_bound_index = index
+        if interleave_bound == -1:
+            # No parallel dimension
+            return
+
+        if interleave_bound % self.pipeline_depth:
+            raise DiagnosticException(
+                "Ranges of interleave factors not implemented yet"
+            )
+
+        old_block = op.body.block
+        new_region = Region(
+            Block(
+                arg_types=(
+                    t
+                    for arg in old_block.args
+                    for t in repeat(arg.type, self.pipeline_depth)
+                )
+            )
+        )
+        with ImplicitBuilder(new_region) as args:
+            # For each interleaved block replica, a mapping from old values to new values
+            value_map: tuple[dict[SSAValue, SSAValue], ...] = tuple(
+                {} for _ in range(self.pipeline_depth)
+            )
+            for arg_index, new_arg in enumerate(args):
+                old_arg = old_block.args[arg_index // self.pipeline_depth]
+                value_map[arg_index % self.pipeline_depth][old_arg] = new_arg
+                new_arg.name_hint = old_arg.name_hint
+            for block_op in old_block.ops:
+                if isinstance(block_op, memref_stream.YieldOp):
+                    memref_stream.YieldOp(
+                        *([vm[arg] for vm in value_map for arg in block_op.arguments])
+                    )
+                else:
+                    for i in range(self.pipeline_depth):
+                        block_op.clone(value_mapper=value_map[i])
+
+        # New maps are the same, except that they have one more dimension and the
+        # dimension that is interleaved is updated to
+        # `dim * self.pipeline_depth + new_dim`.
+        new_indexing_maps = ArrayAttr(
+            AffineMapAttr(
+                m.data.replace_dims_and_symbols(
+                    (
+                        tuple(
+                            AffineExpr.dimension(i)
+                            for i in range(interleave_bound_index)
+                        )
+                        + (
+                            AffineExpr.dimension(interleave_bound_index)
+                            * self.pipeline_depth
+                            + AffineExpr.dimension(m.data.num_dims),
+                        )
+                        + tuple(
+                            AffineExpr.dimension(i)
+                            for i in range(
+                                interleave_bound_index + 1, m.data.num_dims + 2
```

Comment:
Where do these numbers stem from?

Reply:
The idea is that there is an extra dimension added somewhere in the middle of the affine map domain. This is the slice of the domain after the updated value, so it's going from `inserted_index + 1 to old_len + 1` inclusive.